code/modules/xenoarchaeology/xenoartifact_labeler.dm

```diff
@@ -0,0 +1,247 @@
+	return traits
+
+/obj/item/xenoartifact_labeler/proc/look_for(list/place, culprit) //This isn't really needed but, It's easier to use as a function. What does this even do?
+	for(var/X in place) //Using locate breaks this. Sorry.
+		if(X == culprit)
+			. = TRUE
+	return
+
+/obj/item/xenoartifact_labeler/afterattack(atom/target, mob/user)
+	..()
+	var/obj/item/xenoartifact_label/P = create_label(sticker_name)
+	if(!P.afterattack(target, user, TRUE)) //In the circumstance the sticker fails, usually means you're doing something you shouldn't be
+		qdel(P)
+
+///reset all the options	
+/obj/item/xenoartifact_labeler/proc/clear_selection()
+	sticker_name = null
+	info_list = list()
+	sticker_traits = list()
+	activator = list()
+	minor_trait = list()
+	major_trait = list()
+	malfunction = list()
+	ui_update()
+
+/obj/item/xenoartifact_labeler/proc/create_label(new_name)
+	var/obj/item/xenoartifact_label/P = new(get_turf(src))
+	if(new_name)
+		P.name = new_name
+		P.set_name = TRUE
+	P.trait_list = sticker_traits
+	P.info = activator+minor_trait+major_trait+malfunction
+	return P
+
+/obj/item/xenoartifact_labeler/proc/trait_toggle(action, toggle_type, var/list/trait_list, var/list/active_trait_list)
+	var/datum/xenoartifact_trait/description_holder
+	var/new_trait
+	for(var/t in trait_list)
+		new_trait = desc2datum(t)
+		description_holder = new_trait
+		if(action == "assign_[toggle_type]_[t]")
+			if(!look_for(active_trait_list, t))
+				active_trait_list += t
+				info_list += initial(description_holder.label_desc)
+				sticker_traits += new_trait
+			else
+				active_trait_list -= t
+				info_list -= initial(description_holder.label_desc)
+				sticker_traits -= new_trait
+
+/obj/item/xenoartifact_labeler/proc/desc2datum(udesc) //This is just a hacky way of getting the info from a datum using its desc becuase I wrote this last and it's not heartbreaking
+	for(var/t in typesof(/datum/xenoartifact_trait))
+		var/datum/xenoartifact_trait/X = t
+		if((udesc == initial(X.desc))||(udesc == initial(X.label_name)))
+			return t
+	return "[udesc]: There's no known information on [udesc]!."
+
+// Not to be confused with labeler
+/obj/item/xenoartifact_label
+	icon = 'icons/obj/xenoarchaeology/xenoartifact_sticker.dmi'
+	icon_state = "sticker_star"
+	name = "artifact label"
+	desc = "An adhesive label describing the characteristics of a Xenoartifact."
+	var/info = "" 
+	var/set_name = FALSE
+	var/mutable_appearance/sticker_overlay
+	var/list/trait_list = list() //List of traits used to compare and generate modifier.
+	var/obj/item/xenoartifact/xenoa_target
+	///reference for timer
+	var/fall_timer
+
+/obj/item/xenoartifact_label/Initialize()
+	icon_state = "sticker_[pick("star", "box", "tri", "round")]"
+	var/sticker_state = "[icon_state]_small"
+	sticker_overlay = mutable_appearance(icon, sticker_state)
+	sticker_overlay.layer = FLOAT_LAYER
+	sticker_overlay.appearance_flags = RESET_ALPHA
+	sticker_overlay.appearance_flags = RESET_COLOR
+	..()
```

Comment:
This needs to be returned, the value returned by this is important for atom initialization